src/core/public/context/context.ts

```diff
@@ -0,0 +1,222 @@
+ *
+ *
+ * @public
+ */
+export interface ContextContainer<
+  TContext extends {},
+  THandlerReturn,
+  THandlerParameters extends any[] = []
+> {
+  /**
+   * Register a new context provider. Throws an exception if more than one provider is registered for the same context
+   * key.
+   *
+   * @param contextName - The key of the `TContext` object this provider supplies the value for.
+   * @param provider - A {@link ContextProvider} to be called each time a new context is created.
+   * @returns The `ContextContainer` for method chaining.
+   */
+  registerContext<TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, THandlerParameters>
+  ): this;
+
+  /**
+   * Create a new handler function pre-wired to context for the plugin.
+   *
+   * @remarks
+   * This must be called when the handler is registered by the consuming plugin. If this is called later in the
+   * lifecycle it will throw an exception.
+   *
+   * @param handler
+   */
+  createHandler(
+    handler: Handler<TContext, THandlerReturn, THandlerParameters>
+  ): (...rest: THandlerParameters) => Promisify<THandlerReturn>;
+}
+
+/** @internal */
+export class ContextContainerImplementation<
+  TContext extends {},
+  THandlerReturn,
+  THandlerParameters extends any[] = []
+> implements ContextContainer<TContext, THandlerReturn, THandlerParameters> {
+  private currentPlugin?: string;
+
+  /**
+   * Used to map contexts to their providers and associated plugin. In registration order which is tightly coupled to
+   * plugin load order.
+   */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, THandlerParameters>;
+      plugin?: PluginName;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesByPlugin = new Map<PluginName, Array<keyof TContext>>();
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>) {}
+
+  public registerContext = <TContextName extends keyof TContext>(
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, THandlerParameters>
+  ): this => {
+    if (this.contextProviders.has(contextName)) {
+      throw new Error(`Context provider for ${contextName} has already been registered.`);
+    }
+
+    const plugin = this.currentPlugin;
+    this.contextProviders.set(contextName, { provider, plugin });
+
+    if (plugin) {
+      this.contextNamesByPlugin.set(plugin, [
+        ...(this.contextNamesByPlugin.get(plugin) || []),
+        contextName,
+      ]);
+    }
+
+    return this;
+  };
+
+  public createHandler = (handler: Handler<TContext, THandlerReturn, THandlerParameters>) => {
+    const plugin = this.currentPlugin;
+    if (!plugin) {
```

Comment:
does it mean that 
1. core services cannot create handler.
2. plugin that creates a container cannot call `createHandler` within the same lifecycle event, because no one calls `setCurrentPlugin`. yes, now they are in different lifecycles, but the server-side code has a different structure.
?

Reply:
Two great finds, thanks!

1. I really want to be able to throw this exception when plugins might be doing something wrong, but it may not be practical. Only option I can think of is to not throw when we know "core code" is running, so it would only throw if this is called after all plugins have _started_.
2. I'll replace the syncing mechanism with a subscription to a BehaviorSubject that gets passed in at construction. This way it has the most recent value immediately when it's created.

Reply:
>Only option I can think of is to not throw when we know "core code" is running, so it would only throw if this is called after all plugins have started.

I also was thinking what if we use a hybrid of the first and the second approaches? For example, if pass an entity that creates a handler or registers a context, we would have a correct link to a plugin or a core service.
```ts
export type ContextSource = PluginWrapper | CoreService<any, any>;
// context.ts
class ContextContainerImplementation { 
  public createHandler = (
    source: ContextSource,
    handler: Handler<TContext, THandlerReturn, THandlerParameters>
  ) => {
    // here we can check if source PluginWrapper or CoreService and throw otherwise
    const plugin = source instanceof PluginWrapper ? source.name : undefined;
    ...
  }
  public registerContext = <TContextName extends keyof TContext>(
    source: ContextSource,
    contextName: TContextName,
    provider: ContextProvider<TContext, TContextName, THandlerParameters>
  ): this => {
    // here we can check if source PluginWrapper or CoreService and throw otherwise
    const plugin = source instanceof PluginWrapper ? source.name : undefined;
   ...
  }
}
// application_service.tsx
export class ApplicationService {
  setup() {
    return {
      registerApp: (source: ContextSource, app: App) => {
        const boundApp: InternalApp = {
          ...app,
          mountHandler: this.mountContext.createHandler(source, app.mount),
        };
      },
    }
  }
  start(){
    this.mountContext.registerContext(this, 'navigateToApp', () => navigateToApp);
  }
}
```
Just thinking loudly

> I'll replace the syncing mechanism with a subscription to a BehaviorSubject that gets passed in at construction

We don't need a BehaviorSubject, because we don't want a downstream plugin to update a value `subject$.next(42)`.The only thing we need is synchronous access to the current value. I didn't manage to find such API for `Observables`, probably we should add our custom operator BehaviorSubject-like wrapper that can track the current value.

Reply:
I think this option would work for ContextContainers that are created by core, but not for ones created by plugins themselves. This is the problem I ran into to begin with.

For instance, how would we connect the calling plugin to this?

```ts
// plugin_a/public/plugin.tx
class PluginA {
  setup({ context }) {
    this.vizContext = context.createContextContainer();
    return {
      registerVizRenderer(name: string, renderer: (context) => void) {
        // where would source come from???
        this.vizContext.createHandler(source, renderer);
     }
  }
}
```

```ts
// plugin_b/public/plugin.ts

class PluginB {
  setup(core, { plugin_a }) {
    plugin_a.registerVizRenderer('canvas', someRenderFn);
  }
}
```

We could require that PluginB calls with the plugin ID string, but I think this would be brittle and easily break. If you change the plugin ID, you'd have change all context and handler registrations.

Reply:
the first approach from https://github.com/elastic/kibana/pull/41251#discussion_r304818869
I update your example to the case when we operate a Plugin object instead of `id`
```typescript
export type ContextSource = PluginWrapper | CoreService<any, any>;
// context.ts
class ContextContainerImplementation { 
  public createHandler = (
    source: ContextSource,
    handler: Handler<TContext, THandlerReturn, THandlerParameters>
  ) => {
    // here we can check if source PluginWrapper or CoreService and throw otherwise
    const plugin = source instanceof PluginWrapper ? source.name : undefined;
    ...
  }
  public registerContext = <TContextName extends keyof TContext>(
    source: ContextSource,
    contextName: TContextName,
    provider: ContextProvider<TContext, TContextName, THandlerParameters>
  ): this => {
    // here we can check if source PluginWrapper or CoreService and throw otherwise
    const plugin = source instanceof PluginWrapper ? source.name : undefined;
   ...
  }
}

// plugin_a/public/plugin.tx
class PluginA {
  setup({ context }) {
    this.vizContext = context.createContextContainer();
    return {
      registerVizRenderer(source: Source, name: string, renderer: (context) => void) {
        this.vizContext.createHandler(source, renderer);
     }
  }
}

// plugin_b/public/plugin.ts

class PluginB {
  setup(core, { plugin_a }) {
    plugin_a.registerVizRenderer(this, 'canvas', someRenderFn);
  }
}
```
> We could require that PluginB calls with the plugin ID string, but I think this would be brittle and easily break. 

my main concerns are about typos or sniffing a plugin id.

> If you change the plugin ID, you'd have change all context and handler registrations.

I'd say changing `plugin id` is a breaking change in any case, even if we don't relay on `plugin id` in context service directly. All plugins dependent on a plugin have to update:
- manifest file
```diff
- requiredPlugins: ['pluginA']
+ requiredPlugins: ['pluginANew']
```
- contracts
```diff
- public setup(core: CoreSetup, deps: {pluginA: IPluginASetup}) {
+ public setup(core: CoreSetup, deps: {pluginANew: IPluginANewSetup}) {
```

PS. I can't help thinking, do we reinvent [service locator pattern](https://martinfowler.com/articles/injection.html#UsingAServiceLocator)?

Reply:
This would work if we binded `this` to the `PluginWrapper` instance, but that would break accessing any members of the plugin's class (eg. `this.registry`, or `this.myPrivateMethod`). I don't think this is an acceptable approach.

Let's discuss this offline tomorrow.

Reply:
I guess we could use the plugin class instance to map back to the plugin name, though this will require that we load plugin bundles and initialize plugins _before_ `ContextService#setup()` is called. This will require some significant changes since right now no plugin code is ran until after all Core services have ran their `setup` functions.

Reply:
Discussed today offline and decided to make this dead-simple: pass in the pluginId explicitly.

Advangates:
- No magic, cannot break in the async scenarios described in the current revision.
- Does not require re-organizing how plugins are instantiated

Disadvantages:
- Sort of a "special string" that plugins must hardcoded
- If the plugin ID changes, then all of the context and handler registrations need to be updated. We don't expect this to be a common scenario.

Reply:
This implementation has been pushed.